pallets/committee/src/lib.rs

```diff
@@ -289,76 +298,104 @@ decl_module! {
+            match Self::voting(hash) {
+                Some(voting) => Self::vote(origin, hash, voting.index, approve),
+                // NOTE: boxing is necessary or the trait system will throw a fit.
+                None if approve => Self::propose(origin, *call),
+                None => Err(Error::<T, I>::FirstVoteReject.into()),
+            }
         }
 
-        /// Enact the referendum
+        /// Votes `approve`ingly (or not, if `false`)
+        /// on an existing `proposal` given by its hash, `index`.
         ///
         /// # Arguments
-        /// * `id` - Pip Id that need to be enacted
-        #[weight = (T::DbWeight::get().reads_writes(6, 2) + 400_000_000, Operational, Pays::Yes)]
-        pub fn vote_enact_referendum(origin, id: PipId) -> DispatchResult {
-            Self::vote_referendum( origin, id,
-                || {
-                    let aye_call = T::EnactProposalMaker::enact_referendum_call(id);
-                    let nay_call = T::EnactProposalMaker::reject_referendum_call(id);
-                    (aye_call, nay_call)
-                })?;
-            let current_did = Context::current_identity::<Identity<T>>().unwrap_or_default();
-            Self::deposit_event(RawEvent::VoteEnactReferendum(current_did, id));
-            Ok(())
-        }
-
-        /// Reject the referendum
+        /// * `proposal` - A hash of the proposal to be voted on.
+        /// * `index` - The proposal index.
+        /// * `approve` - If `true` than this is a `for` vote, and `against` otherwise.
         ///
-        /// # Arguments
-        /// * `id` - Pip Id that need to be rejected
-        #[weight = (T::DbWeight::get().reads_writes(6, 2) + 400_000_000, Operational, Pays::Yes)]
-        pub fn vote_reject_referendum(origin, id: PipId) -> DispatchResult {
-            Self::vote_referendum( origin, id,
-                || {
-                    let aye_call = T::EnactProposalMaker::enact_referendum_call(id);
-                    let nay_call = T::EnactProposalMaker::reject_referendum_call(id);
-                    (nay_call, aye_call)
-                })?;
-            let current_did = Context::current_identity::<Identity<T>>().unwrap_or_default();
-            Self::deposit_event(RawEvent::VoteRejectReferendum(current_did, id));
+        /// # Errors
+        /// * `BadOrigin`, if the `origin` is not a member of this committee.
+        #[weight = (500_000_000, Operational, Pays::Yes)]
+        pub fn vote(
+            origin,
+            proposal: T::Hash,
+            index: ProposalIndex,
+            approve: bool,
+        ) -> DispatchResult {
+            // 1. Ensure `origin` is a committee member.
+            let did = Self::ensure_is_member(origin)?;
+
+            // 2. Ensure a prior proposal exists and that their indices match.
+            let mut voting = Self::voting(&proposal).ok_or(Error::<T, I>::NoSuchProposal)?;
+            ensure!(voting.index == index, Error::<T, I>::MismatchedVotingIndex);
```

Comment:
Why do we need to take the `index` as input at all?

Reply:
So this is pre-existing logic, so I'm not entirely sure, and I was wondering whether I should remove it or not, but I rationalized its existence as: "a proposal might be made in the future with an identical hash, but different index, and passing the index allows you to be sure its the exact same proposal you're voting on". That might not be valuable though. 🤷

Reply:
As discussed, given you can't have duplicate proposals in the committee (for now ;-)) it looks like we can get rid of index. Will capture this as a separate story.